Port MASTG-TEST-0034: Testing Object Persistence (android) (by @appknox) (#3672)
OWASP/owasp-mstg: tests-beta/android/MASVS-CODE/MASTG-TEST-0x34.md

---
title: Unwanted Object Deserialization Using Serializable
platform: android
id: MASTG-TEST-0x34
type: [static]
weakness: MASWE-0088
profiles: [L1, L2]
---

## Overview

If the app deserializes untrusted data without sufficient validation, it becomes vulnerable to malicious object injection. In Android, data can be passed between components via Intent objects. When an application receives a serialized object within an Intent and deserializes it using `ObjectInputStream.readObject()` without type filtering, a malicious application can send a specially crafted Intent containing a malicious serialized object. This can lead to arbitrary code execution, data tampering, privilege escalation, or denial of service depending on the available gadget chains in the application's classpath.

## Steps

1. Reverse engineer the app (@MASTG-TECH-0013).

> 2. Run static analysis (@MASTG-TECH-0014) using @MASTG-TOOL-0110 on the codebase searching for usages of deserialization.

Reviewer: If the MASTG KNOW document is done, I would just refer to this KNOW document in the test steps and also perform a third step in charge of validating if the deserialization is "unstrusted" (e.g, intents from third-party apps, user input, etc) and that there are not safeguards being applied.

Something like:

- 1. Reverse engineer the app (@MASTG-TECH-0013).
- 2. Run a static analysis (@MASTG-TECH-0014) tool on the reverse engineered app targeting object deserialization APIs and related classes (refer to @MASTG-KNOW-TBD for common deserialization patterns)
- 3. Review the identified call sites to determine the origin of the serialized data and whether any safeguards are applied before the object is used.

## Observation

The output should contain a list of locations where object deserialization is used.

## Evaluation

> The test case fails if the app deserializes data received from untrusted sources (e.g., Intent extras, network input) without proper validation or type filtering.

Reviewer: ```suggestion
The test case fails if the app deserializes data received from untrusted sources (e.g., Intent extras from any other application) without proper validation or type filtering.
```
